### PR TITLE
Add sub-examples functionality

### LINKED ISSUE #4: Numbering and Lettering
Is it possible to use numbers and letters that align as in the picture attached?

example

### PR DESCRIPTION
**Pull request by rwmpelstilzchen**
# Goal

This pull request addresses the issue mentioned in #4 (main issue and comments):

- Referencing was added in #6.
- Sub-examples (including referencing of sub-examples) is added in the current pull request.

# Behaviour

The behaviour of simple examples without sub-examples is unchanged, but a special use of `#gloss` is added to accommodate sub-examples: each sub-example is surrounded with parentheses, which are separated by commas. Referencing sub-examples is done in a similar manner to #6: each sub-example can have `label` and `label-supplement` fields.

# Code

In order to avoid code duplication and make things clearer and simpler, the code responsible to typesetting the internal part of the interlinear glosses is moved to a dedicated function, `#typeset_gloss` (within the main `#gloss` function, until typst/typst#4534 is implemented). For readability the code responsible for each sub-example is bundled within a dedicated function of its own, `#add_subexample`.

It was a true pleasure to work on this package 🙂

### CODE REVIEW

**Comment by neunenak** [comment]
Thanks for the PR! Taking a look now

File: leipzig-gloss.typ [hunk 1]
```diff
@@ -40,79 +40,112 @@
...
-
-    item-spacing: 1em,
     gloss-padding: 2.0em, //TODO document these
     left-padding: 0.5em,
     numbering: false,
     breakable: false,
+    ..args
 ) = {
+    // Typesets the internal part of the interlinear glosses. This function does not deal with the external matters of numbering and labelling.
+    let typeset_gloss(
```
Comment (neunenak) [inline leipzig-gloss.typ]: Typst style is apparently to use `kebab-case` not `snake_case`, and the package submission system now lints for that.

File: leipzig-gloss.typ [hunk 2]
```diff
@@ -128,15 +161,32 @@
     style(styles => {
         block(breakable: breakable)[
             #figure(
-                kind: "ling-example",
-                supplement: label-supplement,
+                kind: "gloss",
```
Comment (neunenak) [inline leipzig-gloss.typ]: Good change

File: documentation.typ [hunk 3]
```diff
@@ -135,6 +135,34 @@ parameters:
 )
 ")
 
+Sub-examples can be achieved by using the same `#gloss` function but surrounding each sub-example with parentheses, separated by commas. A global `header` field for the set can be added.
```
Comment (neunenak) [inline documentation.typ]: I don't think I like trying to re-use the `gloss` function with radically-different argument structures depending on how you want a gloss formatted. Maybe this is a style the typst ecosystem is converging on, I would want to check and see what other typst packages that have complicated functions do. But even so, I think that if it would be hard to write a type signature for a function, it's probably not a great API for a user.

I'm not sure what a good alternative would be. The most straightforward thing would be to define a new function called something like `gloss-block`, that had the signature you propose for `gloss` but without changing the original `gloss` function. I'm not sure I like that either, though, since each gloss subentry is in some sense the same structure as the single entry of a gloss without subentries. It's a bit annoying if there are separate `gloss` and `gloss-block` functions that have very similar (and lengthy) argument lists that need to be kept in sync.

I'll keep thinking about this and maybe see if anyone in the typst chat has a suggestion about what would make for a good API. This package is on a 0.x semver version and makes no claims to stability, so there are no issues with changing the arguments of `gloss` and `numbered-gloss` if it makes sense to do.

File: documentation.typ [hunk 4]
```diff
@@ -195,9 +223,29 @@ References to individual examples can be achieved using the `label` argument and
 
 As we have seen in @sorcerers, […].", addl-bindings: (neg: neg, sg: sg, pl: pl))
 
-Labeling uses the Typst #link("https://typst.app/docs/reference/model/figure/")[figure] document element. The `label-supplement`
-parameter fills in the `suppliment` parameter of a `figure`, which is `[example]` by default.
+Labelling uses the Typst #link("https://typst.app/docs/reference/model/figure/")[figure] document element. The `label-supplement`
```
Comment (neunenak) [inline documentation.typ:286]: typo: "labeling"

File: documentation.typ (hunk 3 above)
Comment (rwmpelstilzchen) [reply documentation.typ]: IMHO it is more elegant, and more straightforward and simple from the user’s perspective, not to have different functions for basically two variants of the same thing. This can be thought of like this: an example with no sub-examples is really an example with a list of sub-examples of size 1, and there is a convention where a list of sub-examples of size 1 is not marked with letters (you don’t have sub-example (a) and that’s it). It’s like a *wug* is a special case of *wugs* where the number of wugs is 1… ;-)

Originally I tried to retain the behaviour of `#gloss` for simplex examples (the only kind before this PR) as it is, but you are totally right: we are on 0.x SemVer after all, so it is acceptable to change the argument structure a bit. So, I think that for the sake of consistency of argument structure we can define `#gloss` to take each sub-example as a dictionary, even if there is only one example (i.e. not treating it as special argument-structure-wise). This way all sub-examples are consistently passed through the [argument sink](https://typst.app/docs/reference/foundations/arguments/#argument-sinks). For example₁, the example₂ from the previous PR would go like this (note the new parentheses):

```typst
#numbered-gloss(
 (
  header: [Middle Welsh; modified from _Grammatical number in Welsh_ (1999) by Silva Nurmio (§~2.1.1)],
  source: ([ac], [ny], [allvs], [y], [dewinyon], [atteb], [idav]),
  morphemes: ([and], [#neg], [be_able.#smallcaps[pret].3#sg], [#smallcaps[def]], [sorcerer.#pl], [answer.#smallcaps[inf]], [to.3#sg.#smallcaps[m]]),
  translation: [and the sorcerers could not answer him],
  label: "sorcerers",
  label-supplement: [Example]
 )
)
```

This is elegant and consistent, and presents only a tiny ‘inconvenience’ in the form of another pair of parentheses (but the way I see it is more convenient — in a deeper sense — to have all calls for the function be of the same kind).

If we agree on this way I will push an additional commit that makes use of the proposed behaviour (updating the documentation accordingly, of course).

File: leipzig-gloss.typ (hunk 2 above)
Comment (rwmpelstilzchen) [reply leipzig-gloss.typ]: 🙂

File: documentation.typ (hunk 3 above)
Comment (neunenak) [reply documentation.typ]: I'm not too bothered by the strategy of forcing the `gloss` function to always accept a list of dictionaries, where the keys of those dictionaries represent the actual meaningful arguments to the function. It's also easy enough to define a new function `single-gloss` (or something along those lines) that just unwraps one layer of dictionary for the common case of only having one example per gloss - that gets rid of the need to type the extra parens all the time unless you're using the functionality.

Still, I like the idea of there being one basic function that generates the "meat" of a gloss - some function with `source`, `morphemes`, and `translation` fields, in other words. And then this function would be optionally wrapped with other grouping functions that would have the effect of "reaching inside" and changing the indentation/numbering scheme of the generated glosses. This feels like it would make for a more ergonomic API. It's also possible that the "right" way to design this API long-term will require new scripting features added to typst that don't currently exist.

For now, I'll say go with the dictionary-list strategy you described above, and add a `single-gloss` function that unwraps the parens in the single-gloss case. That way this code can be merged and it can be useful to people now. I expect that in the future I'll want to change the argument structure of these functions, but that can happen in a future version.

Comment (neunenak) [reply documentation.typ]: I just gave the documentation you linked for `expex` a skim. It looks like they go with the convention of `\ex` for a single example, and `\pex` for a multipart example. Perhaps a reasonable typst equivalent of this API would be functions called `gloss` and `multi-gloss` (and then `numbered-gloss` and `numbered-multi-gloss`, although I'm now wondering if defining `numbered-gloss` in the same namespace is actually a principled choice, especially if the API continues to gain variant functions).

Comment (rwmpelstilzchen) [reply documentation.typ]: > Still, I like the idea of there being one basic function that generates the "meat" of a gloss - some function with source, morphemes, and translation fields, in other words. And then this function would be optionally wrapped with other grouping functions that would have the effect of "reaching inside" and changing the indentation/numbering scheme of the generated glosses. This feels like it would make for a more ergonomic API. It's also possible that the "right" way to design this API long-term will require new scripting features added to typst that don't currently exist.

I totally agree. Moreover, I think this should be clearly communicated by the proper terms. See https://github.com/neunenak/typst-leipzig-glossing/pull/7/commits/fd5048fa6b335d6cef1a3f883d60b8aebacad91f, which separates the two entities you describe and names them `gloss()` and `example()` respectively. I think that’s the best solution for this.

---

BTW, unrelated to the matter in question: many thanks for introducing me to Radicle 👾! It’s brilliant, and I really hope it’ll catch. I for one moved [some of my repos](https://app.radicle.xyz/nodes/seed.radicle.garden/users/did:key:z6MkkshLg6RSzq59qDVN72Tmsxdq7L8iJHvu3FHULA9VkZ9u) to the distributed network. This PR is best to be completed here, but I have some new improvements I will send through Radicle after this one is done.

Comment (rwmpelstilzchen) [reply documentation.typ]: > I just gave the documentation you linked for `expex` a skim. It looks like they go with the convention of `\ex` for a single example, and `\pex` for a multipart example. Perhaps a reasonable typst equivalent of this API would be functions called `gloss` and `multi-gloss` (and then `numbered-gloss` and `numbered-multi-gloss`, although I'm now wondering if defining `numbered-gloss` in the same namespace is actually a principled choice, especially if the API continues to gain variant functions).

In short, I don’t think ExPex’s API is well-designed. (I wrote a [PhD thesis](https://gitlab.com/rwmpelstilzchen/phdthesis) with ExPex, so I’m quite familiar with it. It’s an impressive piece of software, and I’m thankful for the developers and maintainers, but in my opinion things could have been designed better from the end user’s point of view.)

**Comment by neunenak** [comment]
Thanks for the updates, I'm taking another look at this now.

**Comment by neunenak** [comment]
Ok, I think I like this API! I might make some minor tweaks to this in future commits but on the whole this chunk of code looks good, and I'll go ahead and merge.

**Comment by rwmpelstilzchen** [comment]
Wonderful! Thank you 🙂
